PR, Palworld Support (Tested on Minecraft and PalWorld)

ExusAltimus opened the pull request:
Fixes https://github.com/Challengermode/CoreRcon/issues/56

Updated the RCONPacket to use gorcon's implementation (https://github.com/gorcon/rcon/blob/master/packet.go).
This fixed the issues I was having with authentication, I suppose it should work with other servers since gorcon claims support with several.

Also added a parameter to RCON called strictCommandPacketIdMatching (default: true) which when enabled matches the command to the response packet (This was the existing behavior).

When set to false the command instead behaves as a queue and any response will complete the command taskSource.
This is because PalWorld always returns 0 for the response packet id (bad).

ExusAltimus commented:
Just tested on a MC server just and it works.

![image](https://github.com/Challengermode/CoreRcon/assets/6841481/859fa3c0-8f66-4df5-ac89-8bdf0590ba85)

xobust reviewed (changes requested):
Hi thanks for the contribution. 

There was a lot of magic constants in the new implementation of RCON Packet.
I can see that there might be a bug where the buffer size is off by one resulting in a missing null pad. 

For the strictCommandPacketIdMatching, does the server guarantee request response order? Or should the concurrent request be limited to 1. Even though the implementation is flawed I agree we could add this as an option.

ExusAltimus commented on src/CoreRCON/Constants.cs:
```diff
@@ -16,7 +16,12 @@ internal class Constants
+        internal const int PACKET_HEADER_SIZE = 8;
+
+        /// <summary>
+        /// The size of the header of an RCON packet.
```
Yeah there is still a double null termination at the end, I didn't dig too deep into the constants I just copied them directly from gorcon

ExusAltimus commented on src/CoreRCON/Constants.cs:
```diff
@@ -16,7 +16,12 @@ internal class Constants
         /// The size of the header of an RCON packet.
         /// </summary>
-        internal const int PACKET_HEADER_SIZE = 12;
+        internal const int PACKET_HEADER_SIZE = 8;
```
I agree I think the Size, Id and Type constitute the "Packet Header", but im assuming in gorcons implementation they dont consider the size as part of the Header, and use it more as a checksum for the rest of the packet

ExusAltimus replied:
https://developer.valvesoftware.com/wiki/Source_RCON_Protocol#:~:text=The%20Source%20RCON%20Protocol%20is,remote%20console%22%2C%20or%20RCON.
"The packet size field is a 32-bit little endian integer, representing the length of the request in bytes. Note that the packet size field itself is not included when determining the size of the packet, so the value of this field is always 4 less than the packet's actual length. The minimum possible value for packet size is 10:"

ExusAltimus replied:
After consideration I reverted it back and just moved some code around in the RCONPacket.cs

ExusAltimus commented on src/CoreRCON/RCON.cs:
```diff
@@ -367,39 +369,48 @@ public async Task<string> SendCommandAsync(string command, TimeSpan? overrideTim
-                        _incomingBuffer[packet.Id] = body + packet.Body;
-                    }
+                    // Get the most recently sent command
+                    taskSource = _pendingCommands
```
Packets are not gaurenteed to come in order

ExusAltimus replied:
// ensuer mutal execution of SendPacketAsync and RCONPacketReceived
          await _semaphoreSlim.WaitAsync();
          
          It also appears that there is some code that already waits for the response

xobust replied:
Okey just for future reference. 
We disabled concurent requests beacause it was broken in CS2.

ExusAltimus replied:
It can be easily updated by setting the max concurent setting on the _sephamoreSlim (basically a concurrent request limiter). the limit should always be 1 when strictCommandIdMatching is false because we can't gaurentee ordered responses (I tested this with the conctest command)

xobust commented on src/CoreRCON/RCON.cs:
```diff
@@ -367,39 +369,48 @@ public async Task<string> SendCommandAsync(string command, TimeSpan? overrideTim
-                    }
+                    // Get the most recently sent command
+                    taskSource = _pendingCommands
+                        .OrderBy(cmd => cmd.Key)
+                        .Select(cmd => cmd.Value)
+                        .FirstOrDefault();
```
If the server doesn't support multiple packages we could just do first or default here.

ExusAltimus replied:
Updated

xobust commented on src/CoreRCON/RCON.cs:
```diff
@@ -367,39 +369,48 @@ public async Task<string> SendCommandAsync(string command, TimeSpan? overrideTim
                     //Avoid yielding
-                    taskSource.SetResult(packet.Body);
+                    taskSource.SetResult(body ?? string.Empty);
                     _pendingCommands.TryRemove(packet.Id, out _);
+                    _incomingBuffer.Remove(packet.Id);
+                }
+                else
+                {
+                    //Append to previous messages
+                    _incomingBuffer[packet.Id] = body + packet.Body;
                 }
             }
             else
             {
-                _logger?.LogWarning("Received packet with no matching command id: {} body: {}", packet.Id, packet.Body);
+                //Avoid yielding
+                taskSource.SetResult(packet.Body);
```
taskSource can be null here if `strictCommandPacketIdMatching == false`.
Add a null check.

ExusAltimus replied:
Added a gaurd clause, should prevent this

xobust reviewed (commented):
LGTM 👍 

Can merge if your rebase in master and fix the comments.

ExusAltimus commented on src/CoreRCON/RCON.cs:
```diff
@@ -367,39 +369,48 @@ public async Task<string> SendCommandAsync(string command, TimeSpan? overrideTim
+            if (!_pendingCommands.TryGetValue(packet.Id, out taskSource))
             {
-                if (_multiPacket)
+                _logger?.LogWarning("Received packet with no matching command id: {} body: {}", packet.Id, packet.Body);
```
I moved the warning to a guard clause

ExusAltimus commented:
Updated the PR with requested changes, also added autoConnect property because PalWorld RCON implementation sucks so it drops the connection about every 20 seconds when a command isnt sent (I verified it wasn't related to RequestTimeout). PalWorld routinely kills the connection which lead to other issues I encountered such as a deadlock with the authentation task when attempting to use the same RCON instance and reconnect (The reader never flushed). Exposed 2 properties Authenticated and Connected, and a method SetPassword. The new version feels very stable on PalWorld even though its dropping the connection. Tested on minecraft, was able to kill the server, restart it and use the same RCON instance without needing to call ConnectAsync again.

ExusAltimus reviewed (commented):
Updated the PR with requested changes, also added autoConnect property because PalWorld RCON implementation sucks so it drops the connection about every 20 seconds when a command isnt sent (I verified it wasn't related to RequestTimeout). PalWorld routinely kills the connection which lead to other issues I encountered such as a deadlock with the authentation task when attempting to use the same RCON instance and reconnect (The reader never flushed). Exposed 2 properties Authenticated and Connected, and a method SetPassword. The new version feels very stable on PalWorld even though its dropping the connection. Tested on minecraft, was able to kill the server, restart it and use the same RCON instance without needing to call ConnectAsync again.

ExusAltimus commented:
Also forgive me but I dont contribute often, what do you mean by "Can merge if your rebase in master"

xobust commented on src/CoreRCON/Extensions.cs:
```diff
@@ -134,18 +134,16 @@ public static async Task<T> TimeoutAfter<T>(this Task<T> task, TimeSpan? timeout
                     // Operation cancelled
                     throw new TimeoutException();
                 }
-                else
-                {
-                    cts.Cancel();
-                }
 
-                return await task;
```
I don't understand this change,  we shouldn't have to cancel the cts if the delay task is completed.

ExusAltimus replied:
The code on the lefft and right are identical, the else is redundant because the if statement will throw

xobust commented on src/CoreRCON/RCON.cs:
```diff
@@ -175,52 +197,42 @@ await writer.CompleteAsync()
                         byte[] byteArr = buffer.Slice(packetStart, packetEnd).ToArray();
                         RCONPacket packet = RCONPacket.FromBytes(byteArr);
-
-                        if (packet.Type == PacketType.AuthResponse)
```
I really like this change moving authenticate to its own Task awaited on connect 👍

xobust commented on src/CoreRCON/RCON.cs:
```diff
@@ -166,6 +183,11 @@ await writer.FlushAsync()
                 await writer.CompleteAsync()
                     .ConfigureAwait(false);
                 _connected = false;
+                if (_pipeCts != null && !_pipeCts.IsCancellationRequested)
+                {
+                    _pipeCts.Cancel(); // Tell reader to stop waiting
+                    _pipeCts = null;
+                }
```
This should not be needed. 
Did you run into cases where the reader waited after the writer was completed. 
There might be need to run `FlushAsync` here instead.

ExusAltimus replied:
You are correct this code has no function, removing it

xobust reviewed (commented):
Sorry for my late reponse. 

I really like the changes in the PR. 
I am doubtful about the need for the Cancellation token though.

xobust commented:
To make the branch mergeable you need to merge or rebase it with master you can run this command: 
`git pull git@github.com:Challengermode/CoreRcon.git`
And then solve all conflicts with a merge editor like vscode.

ExusAltimus commented:
Thank you, I applied the changes you requested, I also merged changes from main

ExusAltimus commented:
I've been successfully using the RCON for palworld for about a week now, palworld auto disconnects the RCON every 20 seconds but autoconnect lets me reuse the same instance without issues

![image](https://github.com/Challengermode/CoreRcon/assets/6841481/54ef0466-713c-4b88-966e-353c3cae19c8)

xobust reviewed (approved):
LGTM 👍

Trekkan commented:
Can we get this merged?  Thanks for everyone's work!

xobust commented:
Hey sorry for the delay the new version should be up shortly. 
Thanks for your contribution :) 
https://www.nuget.org/packages/CoreRCON/5.4.0
